Remove custom bottom sheet behavior

MiSikora opened the pull request:
## Description

Continuation of Kotlin migration from #44. This PR removes [PCBottomSheetBehavior.java](https://github.com/Automattic/pocket-casts-android/blob/61fe9bf02411dbb158ec838d2fc4834099e99a4b/modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/helper/PCBottomSheetBehavior.java) and uses default behavior built into Material library.

We copied `BottomSheetBehavior` from Material some time ago because it [had issues interacting](https://github.com/Automattic/pocket-casts-android-private/pull/1778) with `ViewPager2`. Since then the library have been updated and I did not discover any issues like between these two components.

`PCBottomSheetBehavior` was [updated](https://github.com/Automattic/pocket-casts-android-private/issues/1768) with [the changes from Material](https://github.com/material-components/material-components-android/commit/6178e7146357bbd602115bcb08194a76592d89c0) to prevent some issues. After that component wasn't updated which means it stayed behind the Material library one. 

At some point `PCBottomSheetBehavior` [it was updated](https://github.com/Automattic/pocket-casts-android-private/pull/1862) to help with dragging issues. After migration I actually encountered a small regression. Especially during the initial flinging up or scrolling up motion. Investigating it made me realize that we have other bugs unrelated to the migration. For example, peek height could be left for a moment above value of `0`. This meant that lifting the finger causes the "up next" sheet to be stuck for a short time and then disappear. I tried to showcase this in the animation below but it is not super obvious as during live interaction with the app.

![screen-20240112-152320-ezgif com-optimize(1)](https://github.com/Automattic/pocket-casts-android/assets/30936061/ddd050ee-234c-4a9c-842e-0cf8eeec2eab)

I tried a lot of different approaches but the simplest one was to remove dragging motion and change it to flinging to open the "up next" bottom sheet. It can still be closed by dragging but cannot be opened this way. Users need to either fling or tap on the top right button. I commented in the code any quirks that I found while working on this feature.

## Testing Instructions

Test `ViewPager2` interaction.

1. Play an episode with longer description and with chapters like [this one](https://pca.st/episode/53cbd091-d83f-4b16-b1af-086e2dc8b9bc).
2. Open up the player.
3. Scroll to `Details`.
4. Scroll vertically through the details.
5. Scroll to `Chapters`.
6. You can also scroll to `Bookmarks` if you have any and test if vertical scrolling works there as well.
7. Scroll back to now playing.

Test interaction between the player bottom sheet and the "up next" bottom sheet. Try to "monkey test" it. Do these things in different order. Play with the functionality and try to break it. Things to test:

* The player should open by tapping on the mini-player.
* The player should open by dragging up the mini-player.
* The player should close by dragging it down by the content.
* The player should close by dragging it down by the top bar handle.
* The player should close by flinging it down.
* The player should close by tapping an arrow in the top left corner.
* The player should close by using the system back button.

---

* "Up next" should open by flinging up on the player.
* "Up next" should open by tapping on the queue in the top right corner of the player.
* "Up next" should close by dragging it down by the content.
* "Up next" should close by dragging it down by the top bar handle.
* "Up next" should close by flinging it down.
* "Up next" should close by tapping `x` in the top left corner.
* "Up next" should close by using the system back button.

## Checklist
- [x] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...
<!-- If this PR does not contain UI changes, ignore these items -->
- [x] with different themes
- [x] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

MiSikora commented:
@ashiagr @mchowning Since this is a small behavior change and I'd like to get input from people who are longer with the app.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

MiSikora commented on modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/PlayerContainerFragment.kt:
```diff
@@ -211,8 +215,10 @@ class PlayerContainerFragment : BaseFragment(), HasBackstack {
     }
 
     fun openUpNext() {
-        val upNextFragment = UpNextFragment.newInstance(source = UpNextSource.PLAYER)
-        (activity as? FragmentHostListener)?.showBottomSheet(upNextFragment)
```
This change was made because as far as I can tell there's no point in creating a new `UpNextFragment` when one is already added to `childFragmentManager` in this fragment. This change speeds up showing the queue when tapping on the top right button as there we just show what already exists.

MiSikora commented on modules/features/player/src/main/res/layout/fragment_player_container.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
-<au.com.shiftyjelly.pocketcasts.player.helper.NestedCoordinatorLayout
```
As far as I could tell and test `NestedCoordinatorLayout` didn't help with anything anymore. It was added in the [mentioned PR](https://github.com/Automattic/pocket-casts-android-private/pull/1862) that was helping with dragging issues but I did not experience any.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ashiagr commented on modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/PlayerContainerFragment.kt:
```diff
@@ -84,16 +84,24 @@ class PlayerContainerFragment : BaseFragment(), HasBackstack {
 
         val binding = binding ?: return
 
```
Since now we always use embedded `UpNextFragment`, can we check use of `embedded` param here:

`val upNextFragment = UpNextFragment.newInstance(embedded = true, source = UpNextSource.NOW_PLAYING)`

It seems to be used in the `UpNextFragment` for styling status and nav bars but I cannot make out any difference  specially after the recent Up Next theming changes in https://github.com/Automattic/pocket-casts-android/pull/1605.

MiSikora replied:
Good point! I'll take a look at it before merging.

ashiagr reviewed (dismissed):
Awesome job! 🙌 

I did a fair amount of testing and did not encounter any noticeable difference for the following scenarios:
- gestures to open/ close player screen
- open/ close upnext screen from full-screen player
- open upnext screen from mini player
- back button handling
- vertical scrolling for large content in Details or Bookmark tab
- scrolling between tabs

I left one comment related to `embedded` param for the upnext screen. Can you retrace its usage and make sure it doesn't cause any styling issues?

MiSikora commented:
@ashiagr

> I left one comment related to `embedded` param for the upnext screen. Can you retrace its usage and make sure it doesn't cause any styling issues?

So this is a bit more complicated than I initially thought. When the `UpNext` is embedded it means that it is treated as a bottom sheet. In consequence this means that the code responsible for settings status and navigation bar colors is this.

https://github.com/Automattic/pocket-casts-android/blob/082d72f501245cc6c3a3cd3f772f1ce0e65f12bc/modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/PlayerContainerFragment.kt#L100-L120

This contains a bug because we always used the dark theme regardless of the user settings. I addressed this is c26a8d23ed56d14a952415b60229f5aca6c9743a.

---

There is also another bug. `BaseFragment` resets navigation and status bars themes in `onResume()` callback.

https://github.com/Automattic/pocket-casts-android/blob/082d72f501245cc6c3a3cd3f772f1ce0e65f12bc/modules/services/views/src/main/java/au/com/shiftyjelly/pocketcasts/views/fragments/BaseFragment.kt#L50-L58

`UpNextFragment` does this in `onAttach()` and this fragment contains custom logic for setting its theme. But those settings are more or less ignored because  `onResume()` executes much later and overrides whatever we wanted to set. I addressed this in 39fe383fda8ac7e8e6f3478237d7bfb62340e901. I'm not 100% happy with this solution but given how fragile and bi-directional theming code and dependencies are within our codebase I think it is ok to do it this way.

---

After those two commits, however the up next queue is opened it should theme device bars correctly. You can test it the following cases.

1. Open the "up next" queue from the main screen by tapping the "up next" queue button on the mini player.
2. Open the "up next" queue from the player screen by tapping the "up next" queue in the top right corner.
3. Open the "up next" queue from the player screen by flinging it up.

If I'm not mistaken those are all entry points for the "up next" queue. This should be tested with different themes and with the `Use Dark Up Next Theme` custom setting.

ashiagr reviewed (approved):
> So this is a bit more complicated than I initially thought. 

I agree, it was not straightforward. Thanks for looking into it 💪

> You can test it the following cases ...

It looks much better to my eyes 🥇 

> If I'm not mistaken those are all entry points for the "up next" queue. 

We can also access UpNext from the app shortcut. I noticed that the status bar does not match with up next top bar through this entry point. The production app also has same style, it is not something introduced in this PR. I feel that it can be tackled separately if we have to. 

<img width=320 src="https://github.com/Automattic/pocket-casts-android/assets/1405144/f4b48002-cc0d-41db-84d9-6a712d3dc389"/>


I'll go ahead and merge this PR. Great work!
